### PR TITLE
Lower intrinsics calls: forget, size_of, unreachable, wrapping_*

### LINKED ISSUE #32716: [MIR] Consider lowering some intrinsic functions to MIR
Some intrinsic functions can be effectively implemented in pure MIR code. Notably, the wrapping and checked arithmetic functions, which can turn into regular `BinaryOp` and `CheckedBinaryOp` (when implemented) rvalues. Similarly intrinsics like `move_val_init` can be exactly represented in the MIR.

While relatively low impact right now, due to most intrinsics being used via wrapper functions, if/when we get MIR-level inlining, this would allow those intrinsics to better participate in optimisations.

### PR DESCRIPTION
**Pull request by tmiasko**
This allows constant propagation to evaluate `size_of` and `wrapping_*`,
and unreachable propagation to propagate a call to `unreachable`.

The lowering is performed as a MIR optimization, rather than during MIR
building to preserve the special status of intrinsics with respect to
unsafety checks and promotion.

Currently enabled by default to determine the performance impact (no
significant impact expected). In practice only useful when combined with
inlining since intrinsics are rarely used directly (with exception of
`unreachable` and `discriminant_value` used by built-in derive macros).

Closes #32716.

### CODE REVIEW

**Review by jonas-schievink — APPROVED** [review]
Changes LGTM, let's see if this impacts perf at all

File: compiler/rustc_mir/src/transform/lower_intrinsics.rs [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                        // during codegen. Issue #35310.
+                    }
+                    sym::size_of => {
+                        if let Some((destination, target)) = *destination {
+                            let tp_ty = substs.type_at(0);
+                            block.statements.push(Statement {
+                                source_info: terminator.source_info,
+                                kind: StatementKind::Assign(box (
+                                    destination,
+                                    Rvalue::NullaryOp(NullOp::SizeOf, tp_ty),
```
Comment (oli-obk) [inline compiler/rustc_mir/src/transform/lower_intrinsics.rs:84]: Hmm... we could also remove this nullary op and replace it with `Rvalue::Use(Operand::Constant(...))`. This applies to all nullary `const fn`.

**Comment by oli-obk** [comment]
cc @rust-lang/wg-mir-opt

**Comment by jonas-schievink** [comment]
Ah, the test output changes on `panic=abort` targets. Maybe try using `-C panic=abort` for that test?

File: compiler/rustc_mir/src/transform/lower_intrinsics.rs (hunk 1 above)
Comment (tmiasko) [reply compiler/rustc_mir/src/transform/lower_intrinsics.rs:84]: We could also teach constant propagation to evalaute rust intrinsics more generally?

**Comment by RalfJung** [comment]
Interesting PR, thanks! Would be nice to Cc @rust-lang/miri for things involving intrinsics as this made some code in Miri (and the CTFE core engine) dead that used to implement those intrinsics.

Also, I expected to see code being removed from the LLVM backend now that it does not have to handle these intrinsics any more; is there any reason this code is kept around?

File: compiler/rustc_mir/src/transform/lower_intrinsics.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/transform/lower_intrinsics.rs:84]: > This applies to all nullary const fn.

Not really, now that CTFE can allocate memory... it only applies to most nullary intrinsics.

File: compiler/rustc_mir/src/transform/mod.rs [hunk 2]
```diff
@@ -390,6 +391,7 @@ fn run_optimization_passes<'tcx>(tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
 
     // The main optimizations that we do on MIR.
     let optimizations: &[&dyn MirPass<'tcx>] = &[
+        &lower_intrinsics::LowerIntrinsics,
```
Comment (RalfJung) [inline compiler/rustc_mir/src/transform/mod.rs:394]: Currently, the optimization is run even with mir-opt-level=0. This means codegen and CTFE can rely on these intrinsics not existing any more. Is that deliberate? For something called "lowering", it makes sense to be able to rely on it, but then it should be moved to `run_post_borrowck_cleanup_passes` IMO.

Comment (tmiasko) [reply compiler/rustc_mir/src/transform/mod.rs:394]: It doesn't run at mir-opt-level=0. The default situation might also change if CTFE is to be run on unoptimized MIR. Always lowering them seems fine to me, modulo concerns already mentioned in the PR description.

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/mod.rs:394]: Are you sure? In https://github.com/rust-lang/rust/pull/79989 I tried `-Zmir-opt-level=0` and it did not make a difference...

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/mod.rs:394]: I see that the pass is only added in `optimizations`, but then CTFE should fail when the `forget` intrinsic is called in https://github.com/rust-lang/rust/pull/79989, and yet I still see this trace:
```
│ │└┐rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
│ │ ├─0ms  INFO rustc_mir::interpret::step // executing bb0
│ │ ├─0ms  INFO rustc_mir::interpret::step _0 = const ()
│ │ ├─0ms  INFO rustc_mir::interpret::step return
│ │ ├─0ms  INFO rustc_mir::interpret::eval_context popping stack frame (returning from function)
│ │┌┘rustc_mir::interpret::eval_context::frame std::mem::forget::<Vec<i32>>
```
Note the lack of anything happening in `mem::forget`. This was with `RUSTC_LOG=rustc_mir::interpret=info rustc +196ca161be8363503d2d2445bab9840f0ce2bba4 /tmp/forget.rs -Zmir-opt-level=0`, on this code:
```rust
fn main() {
    const fn forget_arg_and_return_4(x: Vec<i32>) -> i32 {
        std::mem::forget(x);
        4
    }

    const FOUR_THE_HARD_WAY: i32 = forget_arg_and_return_4(Vec::new());

    assert_eq!(FOUR_THE_HARD_WAY, 4);
}
```

Comment (tmiasko) [reply compiler/rustc_mir/src/transform/mod.rs:394]: Unless you are building std with -Zmir-opt-level=0 it is already too late for it to have an effect for non-local MIR.

Comment (RalfJung) [reply compiler/rustc_mir/src/transform/mod.rs:394]: Oh, right... so we now have a barely tested code path here, namely the intrinsic implementations in CTFE and codegen, which is only hit when std is compiled with mit-opr-level=0. That seems potentially problematic.

**Pull request merged**